### PR TITLE
[pigeon] Started transitioning to run_tests.dart / added a windows support

### PR DESCRIPTION
**Pull request by gaaclarke**
Added a test runner that will generate pigeon code and run the tests that @stuartmorgan setup for windows.

issues:
 * https://github.com/flutter/flutter/issues/91061
 * https://github.com/flutter/flutter/issues/85068

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/pigeon/bin/run_tests.dart [hunk 1]
```diff
@@ -0,0 +1,194 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+////////////////////////////////////////////////////////////////////////////////
+/// Script for executing the Pigeon tests
+///
+/// This currently only supports windows tests.
```
Comment (stuartmorgan) [inline packages/pigeon/bin/run_tests.dart:8]: Nit: Windows

File: packages/pigeon/bin/run_tests.dart [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+      description: 'Compilation tests on generated Dart code.'),
+  'dart_unittests': _TestInfo(
+      function: _runDartUnitTests,
+      description: 'Unit tests on and analysis on Pigeon\'s implementation.'),
+  'flutter_unittests': _TestInfo(
+      function: _runFlutterUnitTests,
+      description: 'Unit tests on generated Dart code.'),
+  'ios_e2e_tests': _TestInfo(
+      function: _runIosE2eTests,
+      description: 'End-to-end objc tests run on iOS Simulator'),
```
Comment (stuartmorgan) [inline packages/pigeon/bin/run_tests.dart:45]: Nit: ObjC (or even better, Objective-C)

File: packages/pigeon/bin/run_tests.dart [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+  'ios_unittests': _TestInfo(
+      function: _runIosUnitTests,
+      description: 'Unit tests on generated Objc code.'),
```
Comment (stuartmorgan) [inline packages/pigeon/bin/run_tests.dart:48]: Same.

File: packages/pigeon/bin/run_tests.dart [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+      cppSourceOut,
+    ]);
+  }
+  if (!hasDart) {
+    args.add('--one_language');
+  }
+  final Process generate = await _streamOutput(Process.start('dart', args));
+  final int generateCode = await generate.exitCode;
+  if (generateCode != 0) {
+    return generateCode;
```
Comment (stuartmorgan) [inline packages/pigeon/bin/run_tests.dart:120]: Shouldn't this print an error message?

File: packages/pigeon/bin/run_tests.dart [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+
+Future<int> _runWindowsUnitTests() async {
+  const String windowsUnitTestsPath = './platform_tests/windows_unit_tests';
+  final int generateCode = await _runPigeon(
+    input: './pigeons/message.dart',
+    cppHeaderOut: '$windowsUnitTestsPath/windows/test/message.g.h',
+    cppSourceOut: '$windowsUnitTestsPath/windows/test/message.g.cpp',
+  );
+  if (generateCode != 0) {
+    return generateCode;
```
Comment (stuartmorgan) [inline packages/pigeon/bin/run_tests.dart:133]: Error messages at the specific failure point would help with understanding/debugging failures.

**Review by stuartmorgan — APPROVED** [review]
Small nits, but LGTM as a starting point.

File: packages/pigeon/bin/run_tests.dart (hunk 4 above)
Comment (gaaclarke) [reply packages/pigeon/bin/run_tests.dart:120]: Nope, if the subprocess returns a nonzero it should have printed out the error already.  I'm streaming stdout and stderr.

File: packages/pigeon/bin/run_tests.dart (hunk 5 above)
Comment (gaaclarke) [reply packages/pigeon/bin/run_tests.dart:133]: Same thing, stdout/stderr streaming should cover that.

File: packages/pigeon/bin/run_tests.dart (hunk 2 above)
Comment (gaaclarke) [reply packages/pigeon/bin/run_tests.dart]: done

File: packages/pigeon/bin/run_tests.dart (hunk 3 above)
Comment (gaaclarke) [reply packages/pigeon/bin/run_tests.dart]: done

File: packages/pigeon/bin/run_tests.dart (hunk 1 above)
Comment (gaaclarke) [reply packages/pigeon/bin/run_tests.dart]: done

**Pull request merged**